Error Reporting

PaulTT88 opened the pull request.

PaulTT88 committed:
Added automated error reporting.

aPunch commented:
This is a good concept, but adding ServerUtils.ErrorReport to any try/catch block is tedious and impractical. We would need to add try/catch blocks to every method. If there is a better way to do this (i.e. a way that doesn't involve adding try/catch blocks to every piece of code) then it would be a great addition and provide much easier error reporting.

PaulTT88 commented:
Agreed

But anywhere where it is possible for an exception to be thrown should be encapsulated in try catch, in my opinion...
It however, would be nicer if there is one place where a try catch block could be placed, like a main program loop, to catch the majority of errors instead of needing to be added to every single block, maybe someone with more java experience would be kind enough to look to see if there is a spot in the code it could be placed.

As more description to what this does, upon exception, it sends the stacktrace to the server, and this will be logged, with statistics of what version, what user, and a small section of syntax highlighted code showing where the exception occured.

Purely for development purposes to streamline bughunting, as not every user is either smart enough to put a stack trace down, or even bothered to report it and continue using citizens.

PaulTT88 commented:
Oops.

PaulTT88 commented on src/core/net/citizensnpcs/Citizens.java:
```diff
@@ -3,7 +3,12 @@ package net.citizensnpcs;
+import java.io.StringWriter;
+import java.io.PrintWriter;
 import java.net.URL;
+import java.net.URLEncoder;
```
I should probably have removed this... And the OutputStreamWriter,StringWriter,PrintWriter and UrlConnection... I originally wrote the Error Reporter in this file then forgot to remove these lines.

metmanr commented:
hey guys i have a problem with this:

Sry i just have no time to write all detailed things down here and I think i won't
use Citizen now because it laggs my server (bad computer ;])
If you fixed this problem pls send me an e-mail on "metmanr@web.de"

Bukkit [1060]
Minecraft 1.7.3

Plugins:

```
AnimalCompanion
AutoRegionGuard
BackToBody
BlockBlocker

Citizens 1.0.9b
Quester
Guard

CraftBukkitUpToDate
CreativeGates
CustomMusic
Factions
FastMining
FootChicken
HeadBlocks
IBIFC (I believe I can fly)
LPMCNavigator
MarcoPolo
mPasswordProtector
mxOnlineTime
NoMoreRain
OwnBlocks
PetCreeper
PTMBukkit (Phoenix Terrain Mod)
SimpleAdmin
Sponge
TimeLapse
WolfPound
WorldEdit
```

This comes up if I click rightclick or leftclick on all my created NPCs:

03:06:29 [INFO] [Citizens] Reloading configuration settings....
03:06:29 [SCHWERWIEGEND] unacceptable character #FFFD special characters are not
 allowed
in "<reader>", position 1164
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.reader.StreamReader.checkP
rintable(StreamReader.java:68)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.reader.StreamReader.update
(StreamReader.java:159)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.reader.StreamReader.peek(S
treamReader.java:117)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.scanner.ScannerImpl.scanPl
ain(ScannerImpl.java:1569)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.scanner.ScannerImpl.fetchP
lain(ScannerImpl.java:860)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.scanner.ScannerImpl.fetchM
oreTokens(ScannerImpl.java:348)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.scanner.ScannerImpl.checkT
oken(ScannerImpl.java:183)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.parser.ParserImpl$ParseBlo
ckMappingKey.produce(ParserImpl.java:564)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.parser.ParserImpl.peekEven
t(ParserImpl.java:163)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.parser.ParserImpl.checkEve
nt(ParserImpl.java:148)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeM
appingNode(Composer.java:228)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeN
ode(Composer.java:160)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeM
appingNode(Composer.java:230)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeN
ode(Composer.java:160)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeM
appingNode(Composer.java:230)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeN
ode(Composer.java:160)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeD
ocument(Composer.java:122)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.getSingl
eNode(Composer.java:105)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.constructor.BaseConstructo
r.getSingleData(BaseConstructor.java:124)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.Yaml.load(Yaml.java:264)
03:06:29 [SCHWERWIEGEND]        at org.bukkit.util.config.Configuration.load(Con
figuration.java:82)
03:06:29 [SCHWERWIEGEND]        at com.citizens.properties.ConfigurationHandler.
load(ConfigurationHandler.java:55)
03:06:29 [SCHWERWIEGEND]        at com.citizens.properties.ConfigurationHandler.
<init>(ConfigurationHandler.java:33)
03:06:29 [SCHWERWIEGEND]        at com.citizens.properties.properties.UtilityPro
perties.initialize(UtilityProperties.java:21)
03:06:29 [SCHWERWIEGEND]        at com.citizens.commands.commands.BasicCommands.
reload(BasicCommands.java:98)
03:06:29 [SCHWERWIEGEND]        at sun.reflect.NativeMethodAccessorImpl.invoke0(
Native Method)
03:06:29 [SCHWERWIEGEND]        at sun.reflect.NativeMethodAccessorImpl.invoke(U
nknown Source)
03:06:29 [SCHWERWIEGEND]        at sun.reflect.DelegatingMethodAccessorImpl.invo
ke(Unknown Source)
03:06:29 [SCHWERWIEGEND]        at java.lang.reflect.Method.invoke(Unknown Sourc
e)
03:06:29 [SCHWERWIEGEND]        at com.citizens.resources.sk89q.CommandsManager.
executeMethod(CommandsManager.java:518)
03:06:29 [SCHWERWIEGEND]        at com.citizens.resources.sk89q.CommandsManager.
execute(CommandsManager.java:404)
03:06:29 [SCHWERWIEGEND]        at com.citizens.Citizens.onCommand(Citizens.java
:196)
03:06:29 [SCHWERWIEGEND]        at org.bukkit.command.PluginCommand.execute(Plug
inCommand.java:35)
03:06:29 [SCHWERWIEGEND]        at org.bukkit.command.SimpleCommandMap.dispatch(
SimpleCommandMap.java:129)
03:06:29 [SCHWERWIEGEND]        at org.bukkit.craftbukkit.CraftServer.dispatchCo
mmand(CraftServer.java:352)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetServerHandler.handleC
ommand(NetServerHandler.java:737)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetServerHandler.chat(Ne
tServerHandler.java:701)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetServerHandler.a(NetSe
rverHandler.java:694)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.Packet3Chat.a(Packet3Cha
t.java:33)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetworkManager.b(Network
Manager.java:226)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetServerHandler.a(NetSe
rverHandler.java:89)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetworkListenThread.a(So
urceFile:105)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.MinecraftServer.h(Minecr
aftServer.java:454)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.MinecraftServer.run(Mine
craftServer.java:363)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.ThreadServerApplication.
run(SourceFile:422)

```
           and this comes up if i try to reload the citizens config-file INGAME with "/reload":
```

03:06:29 [INFO] [Citizens] Reloading configuration settings....
03:06:29 [SCHWERWIEGEND] unacceptable character #FFFD special characters are not
 allowed
in "<reader>", position 1164
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.reader.StreamReader.checkP
rintable(StreamReader.java:68)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.reader.StreamReader.update
(StreamReader.java:159)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.reader.StreamReader.peek(S
treamReader.java:117)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.scanner.ScannerImpl.scanPl
ain(ScannerImpl.java:1569)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.scanner.ScannerImpl.fetchP
lain(ScannerImpl.java:860)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.scanner.ScannerImpl.fetchM
oreTokens(ScannerImpl.java:348)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.scanner.ScannerImpl.checkT
oken(ScannerImpl.java:183)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.parser.ParserImpl$ParseBlo
ckMappingKey.produce(ParserImpl.java:564)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.parser.ParserImpl.peekEven
t(ParserImpl.java:163)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.parser.ParserImpl.checkEve
nt(ParserImpl.java:148)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeM
appingNode(Composer.java:228)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeN
ode(Composer.java:160)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeM
appingNode(Composer.java:230)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeN
ode(Composer.java:160)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeM
appingNode(Composer.java:230)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeN
ode(Composer.java:160)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.composeD
ocument(Composer.java:122)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.composer.Composer.getSingl
eNode(Composer.java:105)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.constructor.BaseConstructo
r.getSingleData(BaseConstructor.java:124)
03:06:29 [SCHWERWIEGEND]        at org.yaml.snakeyaml.Yaml.load(Yaml.java:264)
03:06:29 [SCHWERWIEGEND]        at org.bukkit.util.config.Configuration.load(Con
figuration.java:82)
03:06:29 [SCHWERWIEGEND]        at com.citizens.properties.ConfigurationHandler.
load(ConfigurationHandler.java:55)
03:06:29 [SCHWERWIEGEND]        at com.citizens.properties.ConfigurationHandler.
<init>(ConfigurationHandler.java:33)
03:06:29 [SCHWERWIEGEND]        at com.citizens.properties.properties.UtilityPro
perties.initialize(UtilityProperties.java:21)
03:06:29 [SCHWERWIEGEND]        at com.citizens.commands.commands.BasicCommands.
reload(BasicCommands.java:98)
03:06:29 [SCHWERWIEGEND]        at sun.reflect.NativeMethodAccessorImpl.invoke0(
Native Method)
03:06:29 [SCHWERWIEGEND]        at sun.reflect.NativeMethodAccessorImpl.invoke(U
nknown Source)
03:06:29 [SCHWERWIEGEND]        at sun.reflect.DelegatingMethodAccessorImpl.invo
ke(Unknown Source)
03:06:29 [SCHWERWIEGEND]        at java.lang.reflect.Method.invoke(Unknown Sourc
e)
03:06:29 [SCHWERWIEGEND]        at com.citizens.resources.sk89q.CommandsManager.
executeMethod(CommandsManager.java:518)
03:06:29 [SCHWERWIEGEND]        at com.citizens.resources.sk89q.CommandsManager.
execute(CommandsManager.java:404)
03:06:29 [SCHWERWIEGEND]        at com.citizens.Citizens.onCommand(Citizens.java
:196)
03:06:29 [SCHWERWIEGEND]        at org.bukkit.command.PluginCommand.execute(Plug
inCommand.java:35)
03:06:29 [SCHWERWIEGEND]        at org.bukkit.command.SimpleCommandMap.dispatch(
SimpleCommandMap.java:129)
03:06:29 [SCHWERWIEGEND]        at org.bukkit.craftbukkit.CraftServer.dispatchCo
mmand(CraftServer.java:352)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetServerHandler.handleC
ommand(NetServerHandler.java:737)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetServerHandler.chat(Ne
tServerHandler.java:701)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetServerHandler.a(NetSe
rverHandler.java:694)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.Packet3Chat.a(Packet3Cha
t.java:33)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetworkManager.b(Network
Manager.java:226)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetServerHandler.a(NetSe
rverHandler.java:89)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.NetworkListenThread.a(So
urceFile:105)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.MinecraftServer.h(Minecr
aftServer.java:454)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.MinecraftServer.run(Mine
craftServer.java:363)
03:06:29 [SCHWERWIEGEND]        at net.minecraft.server.ThreadServerApplication.
run(SourceFile:422)

PaulTT88 committed:
Fixed ant builds not showing stack trace information correctly.
Stopped ant complaining about 'includeantruntime'.
Added settings node for turning error reporting on and off.
Added settings node for a custom identification string in error reporting.

PaulTT88 commented:
Tidied the code and fixed the build.xml up a little.

PaulTT88 commented:
## Below is example of data transmitted to server, note it's unknown source because the exception happens in java's own code, and it is run on a machine with only JRE, not JDK, so java's own source is not available for debug info, or at least according to the research I have done this is true, the build.xml was changed to reflect what the ant documentation specified was required to get stack traces to appear correctly, but I do need it confirmed by someone.

_POST[Exception] = java.lang.NumberFormatException: For input string: "LetsMakeAnException"
    at java.lang.NumberFormatException.forInputString(Unknown Source)
    at java.lang.Integer.parseInt(Unknown Source)
    at java.lang.Integer.parseInt(Unknown Source)
    at net.citizensnpcs.Citizens.onEnable(Citizens.java:87)
    at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:126)
    at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:885)
    at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:278)
    at org.bukkit.craftbukkit.CraftServer.loadPlugin(CraftServer.java:178)
    at org.bukkit.craftbukkit.CraftServer.enablePlugins(CraftServer.java:161)
    at net.minecraft.server.MinecraftServer.e(MinecraftServer.java:286)
    at net.minecraft.server.MinecraftServer.a(MinecraftServer.java:273)
    at net.minecraft.server.MinecraftServer.init(MinecraftServer.java:149)
    at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:337)
    at net.minecraft.server.ThreadServerApplication.run(SourceFile:422)

_POST[Version] = devBuild-115

_POST[Ident] = Anonymous

PaulTT88 commented:
http://errorreport.citizensnpcs.net/

Thought I'd give a general idea how the information for received error reports may be layed out...

(For once I'll actually hit the comment button, not that damn comment & close)
